File: solutions/python/spiral-matrix/1/spiral_matrix.py
def spiral_matrix(size):
    matrix = [[0] * size for _ in range(size)]
    counter = 1

    top,bottom,left,right = 0,size-1,0,size-1

    while counter <= size**2:
        for i in range(left,right+1):
            matrix[top][i] = counter
            counter += 1
        top += 1

        for i in range(top,bottom+1):
            matrix[i][right] = counter
            counter += 1
        right -= 1

        if top <= bottom:
            for i in range(right,left-1,-1):
                matrix[bottom][i] = counter
                counter += 1
            bottom -= 1

        if left <= right:
            for i in range(bottom,top-1,-1):
                matrix[i][left] = counter
                counter += 1
            left += 1

    return matrix

File: solutions/cpp/spiral-matrix/1/spiral_matrix.cpp
#include "spiral_matrix.h"

namespace spiral_matrix {

    std::vector<std::vector<uint32_t> > spiral_matrix(const int size) {
        std::vector<std::vector<uint32_t> > matrix(size, std::vector<uint32_t>(size, 0));

        int counter = 1;
        int top = 0, bottom = size - 1, left = 0, right = size - 1;
        while (counter <= size * size) {
            for (int i = left; i <= right; i++) {
                matrix[top][i] = counter++;
            }
            top++;

            for (int i = top; i<= bottom; i++) {
                matrix[i][right] = counter++;
            }
            right--;

            for (int i = right; i >= left; i--) {
                matrix[bottom][i] = counter++;
            }
            bottom--;

            for (int i = bottom; i >= top; i--) {
                matrix[i][left] = counter++;
            }
            left++;
        }

        return matrix;
    }
} // namespace spiral_matrix
